### PR TITLE
fix(mobile): sync issue due to time mismatch

### PR DESCRIPTION
**Pull request by fyfrey**
Currently, the app can get sync issues because of a time mismatch between server and client. This can both occur due to timezone changes or simply wrong clocks.

With this PR, the client retrieves the time for the next sync request from the server, thus solving the issue.

### CODE REVIEW

**Review by jrasm91 — COMMENTED** [review]
Idk if this solves the problem either. The issue is the database may be in a different timezone than the server still. Really we should just send accurate timestamps with timezones

File: server/src/domain/domain.constant.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { extname } from 'node:path';
 import pkg from 'src/../../package.json';
 
 export const AUDIT_LOG_MAX_DURATION = Duration.fromObject({ days: 100 });
+export const AUDIT_LOG_CLEANUP_DURATION = Duration.fromObject({ days: 101 });
```
Comment (jrasm91) [inline server/src/domain/domain.constant.ts:7]: Why is this needed? Is this an attempt to resolve a potential issue with time differences between the server and the database?

Comment (fyfrey) [reply server/src/domain/domain.constant.ts:7]: yeah. it's simply being cautious. keep stuff for 1 day longer than needed

**Comment by fyfrey** [comment]
> Idk if this solves the problem either. The issue is the database may be in a different timezone than the server still. Really we should just send accurate timestamps with timezones

dates between app/server are now always in UTC; the server and datebase use `timestampz` which has timezones.. so it should be good?

File: server/src/domain/domain.constant.ts (hunk 1 above)
Comment (jrasm91) [reply server/src/domain/domain.constant.ts:7]: While that sounds reasonable, I'd much rather be accurate. If we were accurately requesting the change log, this would be unnecessary.

Comment (fyfrey) [reply server/src/domain/domain.constant.ts:7]: true. better safe than sorry for now?

**Comment by jrasm91** [comment]
> > Idk if this solves the problem either. The issue is the database may be in a different timezone than the server still. Really we should just send accurate timestamps with timezones
> 
> dates between app/server are now always in UTC; the server and datebase use `timestampz` which has timezones.. so it should be good?

When we serialize a date and send it over json we're already using the ISO 8601 which is a timezone aware representation of the date and time. If this was passed along correctly to the database everything would be fine, so something is obviously not working correctly. My suspicion is TypeORM is adjusting the time, since JS Dates are only internally stored as numbers and have no sense of timezones. I believe it is offsetting it (incorrectly) to UTC.

https://github.com/typeorm/typeorm/blob/master/src/util/DateUtils.ts#L45-L80.

**Comment by fyfrey** [comment]
> When we serialize a date and send it over json we're already using the ISO 8601 which is a timezone aware representation of the date and time. If this was passed along correctly to the database everything would be fine, so something is obviously not working correctly. My suspicion is TypeORM is adjusting the time, since JS Dates are only internally stored as numbers and have no sense of timezones. I believe it is offsetting it (incorrectly) to UTC.
> 
> https://github.com/typeorm/typeorm/blob/master/src/util/DateUtils.ts#L45-L80.
> 

I see, typeORM doing random stuff again.
It could also use SystemTime instead of UTC (which might be different on the systems....)
I guess this would affect quite a lot of (our) code. Not sure how to fix this. Needs another PR

**Comment by jrasm91** [comment]
A bigger change I am thinking about is switching to luxon date objects, to which you can do with custom type parsers. I wonder if we left it as a raw string if that would fix this endpoint at least.

**Comment by fyfrey** [comment]
We also mount `/etc/localtime:/etc/localtime:ro` for server and microservices, but not for the database container. Maybe this also leads to issues?

**Comment by jrasm91** [comment]
Right if local time is the same it is never an issue. It should not be a requirement that the client, server, and database are on the same timezone though. I'm guessing this can be reproduced by having the client and server on different timezones.

**Comment by alextran1502** [comment]
Hi Fynn, do you think we still need this PR?

**Comment by fyfrey** [comment]
Yeah, it's still an issue. I'll rebase and make some changes.

**Comment by fyfrey** [comment]
rebased and made sure the app always uses the server time when possible

So, this PR guards against sync issues originating from time divergence between client and server, e.g. due clock resets or timezone changes

**Review by jrasm91 — APPROVED** [review]
I'm fine merging it as is. I think we still need to look at how dates get translated between the API body, server, and database, especially when the server and database are in different timezones, but this can be used in the meantime.

File: server/src/domain/audit/audit.service.ts [hunk 2]
```diff
@@ -44,23 +44,24 @@ export class AuditService {
...
     const audits = await this.repository.getAfter(dto.after, {
       ownerId: userId,
       entityType: dto.entityType,
       action: DatabaseAction.DELETE,
     });
 
-    const duration = DateTime.now().diff(DateTime.fromJSDate(dto.after));
-
     return {
+      timeOfRequest: now.toJSDate(),
```
[the remark this answers is not in the document]
Comment (fyfrey) [reply server/src/domain/audit/audit.service.ts:64]: Even better!

**Comment by jrasm91** [comment]
Superseded by #9100 and others
